src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs

```diff
@@ -3,138 +3,591 @@
+            {
+                NetEventSource.Error(instance, $"{instance} Exception while processing event {streamEvent->Type}: {ex}");
+            }
+            return QUIC_STATUS_INTERNAL_ERROR;
+        }
+    }
 
-        public ValueTask WriteAsync(ReadOnlySequence<byte> buffers, CancellationToken cancellationToken = default) => _provider.WriteAsync(buffers, cancellationToken);
+    /// <summary>
+    /// If the read side is not fully consumed, i.e.: <see cref="ReadsClosed"/> is completed and/or <see cref="ReadAsync(Memory{byte}, CancellationToken)"/> returned <c>0</c>,
+    /// dispose will abort the read side with provided <see cref="QuicConnectionOptions.DefaultStreamErrorCode"/>.
+    /// If the write side hasn't been closed, it'll be closed gracefully as if <see cref="CompleteWrites"/> was called.
+    /// Finally, all resources associated with the stream will be released.
+    /// </summary>
+    /// <returns>A task that represents the asynchronous dispose operation.</returns>
+    public override async ValueTask DisposeAsync()
+    {
+        if (Interlocked.Exchange(ref _disposed, 1) != 0)
+        {
+            return;
+        }
 
-        public ValueTask WriteAsync(ReadOnlySequence<byte> buffers, bool endStream, CancellationToken cancellationToken = default) => _provider.WriteAsync(buffers, endStream, cancellationToken);
+        ValueTask valueTask;
 
-        public ValueTask ShutdownCompleted(CancellationToken cancellationToken = default) => _provider.ShutdownCompleted(cancellationToken);
+        // If the stream wasn't started successfully, gracelessly abort it.
+        if (!_startedTcs.IsCompletedSuccessfully)
+        {
+            // Check if the stream has been shut down and if not, shut it down.
+            if (_shutdownTcs.TryInitialize(out valueTask, this))
+            {
+                StreamShutdown(QUIC_STREAM_SHUTDOWN_FLAGS.ABORT | QUIC_STREAM_SHUTDOWN_FLAGS.IMMEDIATE, _defaultErrorCode);
+            }
+        }
+        else
+        {
+            // Abort the read side of the stream if it hasn't been fully consumed.
+            if (_receiveTcs.TrySetException(new QuicOperationAbortedException(), final: true))
+            {
+                StreamShutdown(QUIC_STREAM_SHUTDOWN_FLAGS.ABORT_RECEIVE, _defaultErrorCode);
+            }
+            // Check if the stream has been shut down and if not, shut it down.
+            if (_shutdownTcs.TryInitialize(out valueTask, this))
+            {
+                StreamShutdown(QUIC_STREAM_SHUTDOWN_FLAGS.GRACEFUL, default);
+            }
+        }
 
-        public ValueTask WaitForWriteCompletionAsync(CancellationToken cancellationToken = default) => _provider.WaitForWriteCompletionAsync(cancellationToken);
+        await valueTask.ConfigureAwait(false);
+        _handle.Dispose();
 
-        public void Shutdown() => _provider.Shutdown();
+        // TODO: memory leak if not disposed
+        _sendBuffers.Dispose();
```

Comment:
How do we make sure send buffers are not disposed in the middle of send operation?

Reply:
By awaiting the `valueTask` above which waits on `SHUTDOWN_COMPLETE` which is the last event on the stream and means both side are closed, so no send operation is in progress.

Reply:
Can you add a comment for this please? This is something I am going to forget very soon :D